cs/src/Connections/RelayTunnelConnector.cs

```diff
@@ -0,0 +1,221 @@
+﻿// <copyright file="RelayTunnelConnector.cs" company="Microsoft">
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Ssh;
+using Microsoft.VsSaaS.TunnelService.Contracts;
+
+namespace Microsoft.VsSaaS.TunnelService;
+
+/// <summary>
+/// Tunnel connector that connects to Tunnel Relay service.
+/// </summary>
+internal sealed class RelayTunnelConnector : ITunnelConnector
+{
+    private const int ReconnectAttemptsDoublingDelay = 7; // After the 6th attemt, each next attempt will happen after a delay of 2^7 * 100ms = 12.8s
+    private const int ReconnectInitialDelayMs = 100;
+
+    private readonly IRelayClient relayClient;
+    private readonly Uri relayUri;
+    private string? accessToken;
+
+    public RelayTunnelConnector(IRelayClient relayClient, string relayUri, string? accessToken)
+    {
+        this.relayClient = Requires.NotNull(relayClient, nameof(relayClient));
+        Requires.NotNull(relayClient.StreamFactory, nameof(IRelayClient.StreamFactory));
+        Requires.NotNull(relayClient.WebSocketSubProtocol, nameof(IRelayClient.WebSocketSubProtocol));
+        Requires.NotNull(relayClient.Trace, nameof(IRelayClient.Trace));
+
+        this.relayUri = new Uri(Requires.NotNull(relayUri, nameof(relayUri)), UriKind.Absolute);
+        this.accessToken = accessToken;
+    }
+
+    private TraceSource Trace => this.relayClient.Trace;
+
+    private string Role => this.relayClient.IsTunnelClient ? "client" : "host";
+
+    private string TunnelAccessScope => this.relayClient.IsTunnelClient ? TunnelAccessScopes.Connect : TunnelAccessScopes.Host;
+
+    /// <inheritdoc/>
+    public async Task ConnectSessionAsync(bool isReconnect, CancellationToken cancellation)
+    {
+        int attempDelayMs = ReconnectInitialDelayMs;
+        bool isDelayNeeded;
+
+        object? errorDescription;
+        var disconnectReason = SshDisconnectReason.ConnectionLost;
+        Exception? exception = null;
+        int attempt = 0;
+        bool isTunnelAccessTokenRefreshed = false;
+        string message;
+
+        while (true)
+        {
+            cancellation.ThrowIfCancellationRequested();
+            attempt++;
+            isDelayNeeded = true;
+            Stream? stream = null;
+            errorDescription = null;
+            try
+            {
+                Trace.TraceInformation("Connecting to {0} tunnel relay {1}", Role, this.relayUri.AbsoluteUri);
+
```

Comment:
Also trace the token properties here, using the `TunelAccessTokenProperties` code I just checked in.